ext/vulnsrc/npm/npm.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2019 clair authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package nodejs implements a vulnerability source updater using the NodeJS
+// Security Working Group vulnerabilities database.
+package npm
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path"
+	"strconv"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/coreos/clair/database"
+	"github.com/coreos/clair/ext/versionfmt"
+	"github.com/coreos/clair/ext/versionfmt/semver"
+	"github.com/coreos/clair/ext/vulnsrc"
+	"github.com/coreos/clair/pkg/gitutil"
+)
+
+const (
+	nodejsSecurityWGRepoUrl = "https://github.com/nodejs/security-wg.git"
+	updaterFlag             = "nodejs-securityWGVulnerabilityUpdater"
+)
+
+var ns database.Namespace
+
+type npmSecurityAdvisory struct {
+	ID                 int      `json:"id"`
+	CVEs               []string `json:"cves"`
+	ModuleName         string   `json:"module_name"`
+	Overview           string   `json:"overview"`
+	CVSSScore          float64  `json:"cvss_score"`
+	VulnerableVersions string   `json:"vulnerable_versions"`
+	PatchedVersions    string   `json:"patched_versions"`
+}
+
+func parseNpmVulns(repositoryPath string) ([]database.VulnerabilityWithAffected, error) {
+	files, err := ioutil.ReadDir(path.Join(repositoryPath, "vuln", "npm"))
+	var vulns []database.VulnerabilityWithAffected
+	if err != nil {
+		return vulns, err
+	}
+
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+
+		fileName := path.Join(repositoryPath, "vuln", "npm", file.Name())
+		vulnsInFile, err := parseNpmVuln(fileName)
+		if err != nil {
+			log.WithField("package", "NodeJS").WithField("fileName", fileName).WithError(err).Error("failed to parse vulnerability file")
+			continue
+		}
+
+		vulns = append(vulns, vulnsInFile...)
+	}
+
+	return vulns, nil
+}
+
+func parseNpmVuln(filePath string) (vulns []database.VulnerabilityWithAffected, err error) {
+	var data []byte
```

Comment:
@cnorthwood 

Since this will be ran in a loop. Can we attempt to do this in a streaming fashion instead of loading the full set of bytes into memory ? 

Opening a file descriptor to this file via `os.Open` https://golang.org/pkg/os/#Open. The FD implements io.Reader and io.Writer. You can then use  a json.Decoder for streaming: https://golang.org/pkg/encoding/json/#Decoder

Reply:
I'll have a look at this, but am unlikely to have time to do so for a few weeks now. The JSON files are themselves pretty small so I didn't think the memory overhead a significant issue.

Reply:
@cnorthwood fair enough, the consideration is more for thinking about scale.